Make java.net.Inet6Address constructor arguments more consistent

LeeTibbert opened the pull request:
Whilst adding javalib Datagram support (PR #3614) new contributor RustedBones noticed that the
constructor arguments for class `Inet6Address` were slightly inconsistent with those for `Inet4Address`.
This inconsistency added unnecessary complexity to an already complex trace of the code.   

Future maintainer time should not be spent/wasted on things which can be corrected today.

This PR isolates & implements recommended changes to `Inet6Address` constructor parameters
so that the parameters common to the `InetAddress` parent class and its two subclasses, `Inet4Address`
& `Inet6Address` are now have an identical declaration.

Thank you, RustedBones, for discovering this situation.

LeeTibbert commented:
@RustedBones 

What do you think of the suggested change?   Does it address your concern?   That is my aim.

LeeTibbert commented:
One aspect about the constructor args and the fix that I should probably explain, for comment here, for future understanding, and for me in 3 months...

Scala arrays are mutable. I am _exceedingly_ defensive when coding API methods, so I want to ensure that no
end user can change the `ipAddress` constructor arguments in all three of `Inet*Address`.

The `val ia6ByteArray = in6Addr.getAddress()` change in static `Inet6Address#formatInet6Address()` adds
an allocation and some small runtime cost, but keeps the constructor args parallel and introduces the
cost of capturing the ipaddress only where it is used (instead of at construction).  
It appears that companion object can access variables inside of their companion class but not 
protected constructors.  If anybody knows a better way of doing this, I am open to suggestions.
I never like to give away runtime cycles, but I like mangled public arrays even less.

WojciechMazur commented on javalib/src/main/scala/java/net/Inet6Address.scala:
```diff
@@ -8,11 +8,11 @@ import scalanative.unsigned._
     host: String,
-    scopeId: Int,
-    val zoneIdent: String
+    private val scopeId: Int,
```
If the scopeId is not referenced from other instances, and it seems it is not because it compiled previously, then we can keep it in the original form `copeId: Int`. It would be treated by the compiler similarly to  Scala 2 `private[this]` 
```scala 
scala> class Foo(x: Int){def show(other: Foo) = println(other.x) }
-- [E173] Reference Error: -----------------------------------------------------
1 |class Foo(x: Int){def show(other: Foo) = println(other.x) }
  |                                                 ^^^^^^^
  |   value x cannot be accessed as a member of (other : Foo) from class Foo.
1 error found
                   
```

LeeTibbert commented:
Wojciech,

Thank you for the spot-on review. Sorry my bungled V1 took your time.

The current submission makes the arguments shared by `InetAddress`, `Inet4Address`, and `Inet6Address`
look the same. Two other issues are too big for this PR.

1) If one is looking for consistency, one would think that the constructor args for both `Inet4Address` and
    `Inet6Address` would both be `private`.  Looks like I figured out the `private` part for `Inet6Address` when
    I implemented it and was too ~~chicken~~ hesitant to make `Inet4Address` have a private constructor.
    Doing the latter upsets an apple cart and a whole lot of technical debt from the Harmony days comes
    falling out.  Given a finite amount of developer time, I suspect that we are just going to have to live
    with the inconsistency for a long time.  Someday, there will be an Issue created describing how
    "new Inet4Address(args)" works and should not and this punt will come back to haunt me.

2) I created Issue #3657 to capture some NetworkInterface related Inet6Address technical debt.
    The affects the 4th (zoneId) constructor argument, so I did not touch that arg in this PR.
    This stuff is a tangle and I would like to keep a somewhat understandable audit trail.

LeeTibbert commented:
Future reference:  for anyone, including me, trying to resolve inconsistency and/or unwind the reasoning.

I implemented the review suggestion (thank you, W.) to make the instance secondary constructors
`private[net]`. They are used only in the superclass `InetAddress`.

Deleted one totally unused secondary constructor to reduce distraction.  In a less constrained world,
I would have moved the two secondary constructors to private `apply` methods in the companion object.
Then all instantiations would go through the companion object instead of some being `apply` and some being
`new`.  I think the general direction of Scala is to use the `apply()` and reduce the need for the `new keyword`.

I left the two constructors as secondary so that the code in `InetAddress` constructs Inet4Address and Inet6Address
instances the same way (using `new`).  When/if Inet4Address gets a private constructor, similar to current Inet6Address,
the code in both can be made more consistent (i.e. both use suitable apply() methods).

Enough shuffling of the deck chairs.

WojciechMazur commented:
> Then with 3 segfaults, probably due to trying to access address 0 (from unresolved link)

Actually this might be due to same rare failures in the GC, hopefully fixed in #3658
The unresolved links are due to changes in debug metadata codegen, they should be harmless. I'll try to fix them
